norduser/service/fork.go

```diff
@@ -3,57 +3,135 @@ package service
 import (
 	"errors"
 	"fmt"
+	"log"
 	"os/exec"
+	"strconv"
 	"strings"
+	"sync"
 	"syscall"
 
 	"github.com/NordSecurity/nordvpn-linux/internal"
-	"golang.org/x/sys/unix"
 )
 
 // ErrNotStarted when disabling norduser
 var ErrNotStarted = errors.New("norduserd wasn't started")
 
 // ChildProcessNorduser manages norduser service through exec.Command
 type ChildProcessNorduser struct {
-	commandHandles map[uint32]*exec.Cmd
+	mu sync.Mutex
 }
 
 func NewChildProcessNorduser() *ChildProcessNorduser {
-	return &ChildProcessNorduser{
-		commandHandles: make(map[uint32]*exec.Cmd),
+	return &ChildProcessNorduser{}
+}
+
+// handlePsError returns nil if err is nil or error code 1(no processes listed). It returns unmodified err in any other
+// case.
+func handlePsError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var exiterr *exec.ExitError
+	if errors.As(err, &exiterr) {
+		// ps returns 1 when no processes are shown
+		if exiterr.ExitCode() == 1 {
+			return nil
+		}
 	}
+
+	return err
 }
 
-func isRunning(uid uint32) (bool, error) {
-	// list all norduserd processes, restrict output to uid of the owner
+func parseNorduserPIDs(psOutput string) []int {
+	pids := []int{}
+	for _, pidStr := range strings.Split(psOutput, "\n") {
+		pid, err := strconv.Atoi(strings.TrimSpace(pidStr))
+		if err != nil {
+			log.Println("failed to parse pid string: ", err)
+			continue
+		}
+
+		pids = append(pids, pid)
+	}
+
+	return pids
+}
+
+func getRunningNorduserPIDs() ([]int, error) {
 	// #nosec G204 -- arguments are constant
-	output, err := exec.Command("ps", "-C", internal.Norduserd, "-o", "uid=").CombinedOutput()
-	if err != nil {
-		var exiterr *exec.ExitError
-		if errors.As(err, &exiterr) {
-			// ps returns 1 when no processes are shown
-			if exiterr.ExitCode() == 1 {
-				return false, nil
-			}
+	output, err := exec.Command("ps", "-C", internal.Norduserd, "-o", "pid=").CombinedOutput()
+	if err := handlePsError(err); err != nil {
+		return []int{}, fmt.Errorf("listing norduser pids: %w", err)
+	}
+
+	return parseNorduserPIDs(string(output)), nil
+}
+
+func findPIDOfUID(uids string, uid uint32) int {
+	desiredUID := fmt.Sprint(uid)
+	for _, pidUid := range strings.Split(uids, "\n") {
+		pidUidSplit := strings.Split(strings.TrimSpace(pidUid), " ")
```

Comment:
[minor] a possible alternative would be to use `Sscanf` to read the 2 ints from the string? and then you wouldn't need to worry about spaces.